tools/cell_census_builder/validate.py

```diff
@@ -271,39 +276,89 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
+            assert ad.n_obs == len(obs_joinids)
+            if len(obs_joinids) == 0:
+                continue
 
-            def count_nonzero(arr: Union[sparse.spmatrix, npt.NDArray[Any]]) -> int:
-                """Return _actual_ non-zero count, NOT the stored value count."""
-                if isinstance(arr, (sparse.spmatrix, sparse.coo_array, sparse.csr_array, sparse.csc_array)):
-                    return np.count_nonzero(arr.data)
-                return np.count_nonzero(arr)
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
+                .X["raw"]
+                .read((obs_joinids.soma_joinid.to_numpy(), slice(None)))
+                .tables()
+                .concat()
+            )
+            data = X_raw["soma_data"].to_numpy()
+            rows = X_raw["soma_dim_0"].to_numpy()
+            cols = X_raw["soma_dim_1"].to_numpy()
+
+            # positionally (re)index obs/rows. We _know_ that the Census assigns
+            # obs soma_joinids in the obs position order of the original AnnData, so
+            # leverage that for simplicity and speed.
+            rows_by_position = obs_joinids.set_index("soma_joinid").index.get_indexer(rows)
+
+            # get the joinids for the var axis
+            all_var_ids = (
+                exp.ms[MEASUREMENT_RNA_NAME].var.read(column_names=["soma_joinid", "feature_id"]).concat().to_pandas()
+            )
+            # mask defines which feature_ids are in the AnnData
+            var_joinid_in_adata = all_var_ids.feature_id.isin(expected_ad_var.index)
+            assert ad.n_vars == var_joinid_in_adata.sum()
+
+            # positionally re-index the vars/cols using the feature_id as the join key.
+            cols_by_position = (
+                expected_ad_var.join(all_var_ids.set_index("feature_id"))
+                .set_index("soma_joinid")
+                .index.get_indexer(cols)
+            )
 
-            if ad.raw is None:
-                assert raw_nnz == count_nonzero(
-                    ad.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.X)}"
-            else:
-                assert raw_nnz == count_nonzero(
-                    ad.raw.X
-                ), f"{eb.name}:{dataset.dataset_id} 'raw' nnz mismatch {raw_nnz} vs {count_nonzero(ad.raw.X)}"
+            # Assertion 1 - the contents of the X matrix are EQUAL for all var values
+            # present in the AnnData
+            assert (
+                sparse.coo_matrix((data, (rows_by_position, cols_by_position)), shape=expected_ad_x.shape)
+                != expected_ad_x
+            ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
+
+            # Assertion 2 - the contents of the X matrix are EMPTY for all var ids
+            # NOT present in the AnnData. Test by asserting that no col IDs contain
+            # a joinid not in the AnnData.
```

Comment:
I'm probably missing something here, since I'm not seeing the case where assertion 1 (full value quality) passes and assertion 2 fails. In other words, if X contained data for an "extra" gene wouldn't that cause the full coo_matrix inequality check, above, to fail? Doesn't the inequality check consider the coords?

E.g, this case, where `ccX` has a value at col1, and `adX` does not, fails the original equality check:
```
In [1]: ccX = sc.sparse.coo_matrix(([0, 1, 2], ([0, 0, 0], [0, 1, 2])), shape=(1,4))

In [2]: adX = sc.sparse.coo_matrix(([0, 2], ([0, 0], [0, 2])), shape=(1,4))

In [3]: (ccX != adX).nnz
Out[3]: 1
```

Reply:
Your assumption fails to catch the case where the Census incorrectly has values in a var (feature_id) which is not present in the AnnData.

The first test restricts itself to the feature_ids present in the AnnData, and confirms that the values and coordinates in the Census are "eq".  But this leaves it blind to the case where the census has garbage values outside of those feature ids.

This approach is used because we need to reindex the matrix using the AnnData var dataframe index (i.e., using it's ordering/iloc).  As that index does not contain var feature_ids not present in itself (:-)), it was simplest to split this test into two separate assertions.